sensor/intrinsic_camera_calibrator/intrinsic_camera_calibrator/intrinsic_camera_calibrator/camera_calibrator.py

```diff
@@ -0,0 +1,1094 @@
+            ImageViewMode.SOURCE_UNRECTIFIED,
+            ImageViewMode.TRAINING_DB_UNRECTIFIED,
+            ImageViewMode.EVALUATION_DB_UNRECTIFIED,
+        }:
+            img = copy.deepcopy(self.unprocessed_image)
+        elif self.image_view_type_combobox.currentData() == ImageViewMode.SOURCE_RECTIFIED:
+            assert self.calibrated_camera_model is not None
+            img = self.calibrated_camera_model.rectify(
+                self.unprocessed_image, self.undistortion_alpha_spinbox.value()
+            )
+        else:
+            raise NotImplementedError
+
+        self.pending_detection_request = False
+        self.pending_detection_result = True
+
+        self.request_image_detection.emit(img)
+
+    def process_db_data(self, img):
+
+        assert self.image_view_type_combobox.currentData() in set(
+            {ImageViewMode.TRAINING_DB_UNRECTIFIED, ImageViewMode.EVALUATION_DB_UNRECTIFIED}
+        )
+
+        with self.lock:
+            self.unprocessed_image = img
+
+        if self.pending_detection_result:
+            self.pending_detection_request = True
+        else:
+            self.should_process_image.emit()
+
+    def process_new_data(self):
+        """Attempt to request the detector to process an image. However, if it there is an image being processed, does not enqueue them indefinitely. Istead, only leave the last one."""
+        if self.paused:
+            return
+
+        if self.image_view_type_combobox.currentData() not in set(
+            {ImageViewMode.SOURCE_RECTIFIED, ImageViewMode.SOURCE_UNRECTIFIED}
+        ):
+            return
+
+        with self.lock:
+            if self.produced_image is not None:
+                self.unprocessed_image = self.produced_image
+            else:
+                self.unprocessed_image = self.produced_image
+                self.produced_image = None
+
+        if self.pending_detection_result:
+            self.pending_detection_request = True
+        else:
+            self.should_process_image.emit()
+
+    def data_source_external_callback(self, img: np.array):
+        """
+        Consumer side of the producer/consumer pattern.
+
+        The producer generally has its own thread so synchronization it is neeeded
+        Args:
+            img (np.array): the produced image coming from any data source.
+        """
+        # We depcouple the the data coming from the source with the processing slot to avoid dropping frames in case we cann not process them all
+        with self.lock:
+            self.produced_image = img
+            self.produced_data_signal.emit()  # Using a signal from another thread results in the slot being executed in the class Qt thread
+
+    def on_parameter_changed(self):
+        self.should_process_image.emit()
+
+
+def main(args=None):
+    app = QApplication(sys.argv)
+
+    # rclpy.init(args=args)
+
+    cfg = {}
+    try:
+        with open(sys.argv[1], "r") as stream:
```

Comment:
use rosparam instead of sys.argv?

Reply:
The node is instantiated only in the case of `ros topic mode`, so it can not be used here

Reply:
Actually, this tools can be ran without ROS :)

Reply:
I see, good point. So handle arguments with optparse style?

Reply:
Fixed in [b894e67](https://github.com/tier4/CalibrationTools/pull/65/commits/b894e67842913ce0ff932710011a6987a4041727)

Reply:
arguments error with using launch file?
```
root@ros-desktop:/workspace# ros2 launch intrinsic_camera_calibrator calibrator.launch.xml 
[INFO] [launch]: All log files can be found below /root/.ros/log/2023-01-26-04-10-48-042498-ros-desktop-220
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [camera_calibrator-1]: process started with pid [223]
[camera_calibrator-1] Usage: camera_calibrator [options]
[camera_calibrator-1] 
[camera_calibrator-1] camera_calibrator: error: incorrect number of arguments: 2
[ERROR] [camera_calibrator-1]: process has died [pid 223, exit code 2, cmd '/workspace/install/intrinsic_camera_calibrator/lib/intrinsic_camera_calibrator/camera_calibrator /workspace/install/intrinsic_camera_calibrator/share/intrinsic_camera_calibrator/config/intrinsics_calibrator.yaml --ros-args -r __node:=camera_calibrator'].
```

Reply:
ros2 run is working.
```
root@ros-desktop:/workspace# ros2 run intrinsic_camera_calibrator camera_calibrator -c install/intrinsic_camera_calibrator/share/intrinsic_camera_calibrator/config/intrinsics_calibrator.yaml 
```

Reply:
[nits] just about style, `--config_file` -> `--config-file`

Reply:
Addressed the nit in [510247c](https://github.com/tier4/CalibrationTools/pull/65/commits/510247ce764001f296c74a1dddb1c9a1ebb4a715)

Reply:
Maybe the launcher has not been updated. Can you try rebuilding with colcon just in case?

Reply:
works well after clean build, sorry! :cry: